File: le1/avl.c
#include "avl.h"
#include <stdlib.h>


// Prototypes
AVL *rebalance(AVL *avl);
AVL *rotate_left(AVL *avl);
AVL *rotate_right(AVL *avl);

int max(int a, int b);
int height(AVL *avl);
void update_height(AVL *avl);
// ===========


struct AVL {
    /**
     * An AVL tree is a binary search tree implementation that "rotates"
     * itself if its two subtrees are unbalanced after any operation to
     * ensure that it remains balanced.
     * 
     * Given an AVL with n nodes, these are its operations and their time complexities:
     *     avl_search: O(log n)
     *     avl_insert: O(log n)
     *     avl_delete: O(log n)
     */
    int64_t value;
    int height;
    struct AVL *left;
    struct AVL *right;
};


AVL *avl_init(int64_t value) {
    /**
     * Initializes an AVL node with the given value, initially with a
     * height of 0 and no left or right child.
     * 
     * Args:
     *     value (int64_t): The value the node will have.
     * 
     * Returns:
     *     AVL*: The initialized AVL node.
     */
    AVL *avl = (AVL*)malloc(sizeof(AVL));
    avl->value = value;
    avl->height = 0;
    avl->left = NULL;
    avl->right = NULL;
    return avl;
}


AVL *avl_search(AVL *avl, int64_t value) {
    /**
     * Searches for the node containing the given value in the AVL
     * tree and returns it if it exists. If it doesn't NULL is returned.
     * 
     * Args:
     *     avl (AVL*): The AVL tree where the value is searched in.
     *     value (int64_t): The value to search for.
     * 
     * Returns:
     *     AVL*: The node containing the given value.
     */
    if (avl == NULL) {
        return NULL;
    } else if (value < avl->value) {
        return avl_search(avl->left, value);
    } else if (value > avl->value) {
        return avl_search(avl->right, value);
    } else {
        return avl;
    }
}


AVL *avl_insert(AVL *avl, int64_t value) {
    /**
     * Inserts the given value into the AVL tree, then rebalances it to
     * maintain the AVL property.
     * =
     * Args:
     *     avl (AVL*): The AVL tree where the value is inserted in.
     *     value (int64_t): The value to insert.
     * 
     * Returns:
     *     AVL*: The new AVL tree post-insertion.
     */
    if (avl == NULL) {
        return avl_init(value);
    } else if (value < avl->value) {
        avl->left = avl_insert(avl->left, value);
    } else if (value > avl->value) {
        avl->right = avl_insert(avl->right, value);
    } else {
        // Do nothing
    }

    update_height(avl);
    return rebalance(avl);
}


AVL *avl_delete(AVL *avl, int64_t value) {
    /**
     * Deletes the given value from the AVL tree, then rebalances it to
     * maintain the AVL property.
     * =
     * Args:
     *     avl (AVL*): The AVL tree where the value is deleted from.
     *     value (int64_t): The value to delete.
     * 
     * Returns:
     *     AVL*: The new AVL tree post-deletion.
     */
    if (avl == NULL) {
        return NULL;
    } else if (value < avl->value) {
        avl->left = avl_delete(avl->left, value);
    } else if (value > avl->value) {
        avl->right = avl_delete(avl->right, value);
    } else {
        // Search for new root
        if (avl->left == NULL) {
            // No left subtree: Use right subtree as new root
            AVL *right = avl->right;
            free(avl);
            return right;
        } else if (avl->right == NULL) {
            // No right subtree: Use left subtree as new root
            AVL *left = avl->left;
            free(avl);
            return left;
        } else {
            // Both subtrees exist: Get leftmost node of right subtree
            AVL *new_root = avl->right;
            while (new_root->left != NULL) {
                new_root = new_root->left;
            }
            avl->value = new_root->value;
            avl->right = avl_delete(avl->right, new_root->value);
        }
    }

    update_height(avl);
    return rebalance(avl);
}


void avl_free(AVL *avl) {
    /**
     * Frees the memory of the given AVL tree recursively.
     * 
     * Args:
     *     avl (AVL*): The AVL tree to free.
     */
    if (avl == NULL) {
        return;
    }
    avl_free(avl->left);
    avl_free(avl->right);
    free(avl);
}


AVL *rebalance(AVL *avl) {
    /**
     * Checks if the AVL tree is balanced, that is, the height difference of
     * its subtrees is at most 1. If the tree is not balanced, it is rotated
     * accordingly. Does nothing if NULL is given.
     * 
     * Args:
     *     avl (AVL*): The AVL tree to rebalance.
     * 
     * Returns:
     *     AVL*: The rebalaned AVL tree.
     * 
     */
    if (avl != NULL) {
        update_height(avl); // Just to be safe
        if (height(avl->left) - height(avl->right) >= 2) {
            // Left subtree too tall: Rotate right
            if (height(avl->left->left) < height(avl->left->right)) {
                // avl->left->left shouldn't be shorter than avl->left->right before the rotation
                avl->left = rotate_left(avl->left);
            }
            avl = rotate_right(avl);
        } else if (height(avl->right) - height(avl->left) >= 2) {
            // Right subtree too tall: Rotate left
            if (height(avl->right->right) < height(avl->right->left)) {
                // avl->right->right shouldn't be shorter than avl->right->left before the rotation
                avl->right = rotate_right(avl->right);
            }
            avl = rotate_left(avl);
        }
    }

    return avl;
}


AVL *rotate_left(AVL *avl) {
    /**
     * Rotates the AVL tree left (counter-clockwise).
     * 
     * Args:
     *     avl (AVL*): The AVL tree to rotate.
     * 
     * Returns:
     *     AVL*: The rotated AVL tree.
     */
    AVL *right = avl->right;
    AVL *right_left = avl->right->left;

    right->left = avl;
    avl->right = right_left;

    update_height(avl);
    update_height(right);

    return right;
}


AVL *rotate_right(AVL *avl) {
    /**
     * Rotates the AVL tree right (clockwise).
     * 
     * Args:
     *     avl (AVL*): The AVL tree to rotate.
     * 
     * Returns:
     *     AVL*: The rotated AVL tree.
     */
    AVL *left = avl->left;
    AVL *left_right = avl->left->right;

    left->right = avl;
    avl->left = left_right;

    update_height(avl);
    update_height(left);

    return left;
}


int max(int a, int b) {
    /**
     * Helper C function for returning the greater number between two numbers.
     * 
     * Args:
     *     a (int): First number.
     *     b (int): Second number.
     * 
     * Returns:
     *     int: The greater of a and b.
     */
    return a > b ? a : b;
}


int height(AVL *avl) {
    /**
     * Helper function for returning the height of an AVL tree, allowing for
     * NULL to be evaluated to -1 instead of having to check for NULL before
     * using avl->height.
     * 
     * Args:
     *     avl (AVL*): The AVL tree whose height to check.
     * 
     * Returns:
     *     int: The height of the given AVL tree, or -1 if NULL was given.
     */
    return avl == NULL ? -1 : avl->height;
}


void update_height(AVL *avl) {
    /**
     * Updates the height property of the given AVL node. Note that it does not
     * recursivley update its children's heights.
     * 
     * Args:
     *     avl (AVL*): The node whose height is to be updated.
     */
    avl->height = max(height(avl->left), height(avl->right)) + 1;
}

File: le1/avl.py
class AVLTree:
    """
    An AVL tree is a self-balancing binary search tree implementation that checks
    the heights of its subtrees and rotates accordingly if imbalanced after every
    operation to ensure that the tree remains balanced.

    Given a treap with n nodes, here are its operations' time complexities:
        - contains: O(log n)
        - add: O(log n)
        - remove: O(log n)
    """
    def __init__(self) -> None:
        self._root: _AVLNode | None = None
        self._size = 0
    
    @property
    def size(self) -> int:
        """The number of values currently in the AVL tree."""
        return self.size

    def contains(self, value: int) -> bool:
        """
        Returns `True` if `value` exists in the AVL tree, and `False` otherwise. If the
        AVL tree is empty, `False` is returned automatically

        Parameters:
            value (int): The value to search for.
        
        Returns:
            bool: Whether the value exists in the treap.
        """
        return _AVLNode.search(self._root, value) is not None

    def add(self, value: int) -> bool:
        """
        Inserts a new value into the AVL tree if it doesn't exist.

        Parameters:
            value (int): The value to insert.
        
        Returns:
            bool: `True` if the value was successfully inserted, `False` if it already exists.
        """
        self._root, success = _AVLNode.insert(self._root, value)
        if success:
            self._size += 1
        return success
    
    def remove(self, value: int) -> bool:
        """
        Removes a value from the AVL tree if it exists.

        Parameters:
            value (int): The value to remove.
        
        Returns:
            bool: True if the value was successfully removed, False if it was not found.
        """
        self._root, success = _AVLNode.delete(self._root, value)
        if success:
            self._size -= 1
        return success


class _AVLNode:
    """
    Internal node class used by the AVLTree. Each node stores:
        - a value (int)
        - its height (int)
        - references to left and right child nodes

    The AVL tree maintains the binary search tree property (ordered by value)
    and ensures balance by checking subtree heights and performing rotations.
    """
    def __init__(self, value: int) -> None:
        self._value = value
        self._height = 0
        self._left: _AVLNode | None = None
        self._right: _AVLNode | None = None
    
    @property
    def value(self) -> int:
        """The value stored in this node."""
        return self.value
    
    @staticmethod
    def search(node: _AVLNode | None, value: int) -> _AVLNode | None:
        """
        Searches for a value in the AVL tree starting from the given node.

        Parameters:
            node (_AVLNode | None): The root of the subtree to search.
            value (int): The value to search for.
        
        Returns:
            _AVLNode|None: The node containing the value, or `None` if not found.
        """
        if node is None:
            return None
        elif value < node.value:
            return _AVLNode.search(node._left, value)
        elif value > node.value:
            return _AVLNode.search(node._right, value)
        else:
            return node
    
    @staticmethod
    def insert(node: _AVLNode | None, value: int) -> tuple[_AVLNode | None, bool]:
        """
        Inserts a new value into the AVL tree rooted at `node`.

        Parameters:
            node (_AVLNode | None): The root of the subtree.
            value (int): The value to insert.
        
        Returns:
            tuple[_AVLNode|None,bool]:
                - The new root of the subtree.
                - Whether insertion succeeded.
        """
        if node is None:
            return _AVLNode(value), True
        elif value < node.value:
            node._left, success = _AVLNode.insert(node._left, value)
        elif value > node.value:
            node._right, success = _AVLNode.insert(node._right, value)
        else:
            return node, False
        
        node = _AVLNode._rebalance(node)
        return node, success
    
    @staticmethod
    def delete(node: _AVLNode | None, value: int) -> tuple[_AVLNode | None, bool]:
        """
        Deletes a value from the AVL tree rooted at `node`.

        Parameters:
            node (_AVLNode | None): The root of the subtree.
            value (int): The value to delete.
        
        Returns:
            tuple[_AVLNode|None,bool]:
                - The new root of the subtree.
                - Whether insertion succeeded.
        """
        if node is None:
            return None, False
        elif value < node.value:
            node._left, success = _AVLNode.delete(node._left, value)
        elif value > node.value:
            node._right, success = _AVLNode.delete(node._right, value)
        else:
            success = True
            if node._left is None or node._right is None:
                return node._left or node._right, True
            else:
                successor = node._right
                while successor._left:
                    successor = successor._left
                node._value = successor.value
                node._right, _ = _AVLNode.delete(node._right, successor.value)
        
        node = _AVLNode._rebalance(node)
        return node, success
    
    @staticmethod
    def height(node: _AVLNode | None) -> int:
        """Returns the height of the given node, or -1 if `None` was given."""
        return node._height if node is not None else -1
    
    @staticmethod
    def _rebalance(node: _AVLNode) -> _AVLNode:
        """
        Rebalances the AVL tree rooted at `node` if it is imbalanced.

        Parameters:
            node (_AVLNode): The root of the subtree.
        
        Returns:
            _AVLNode: The new root of the subtree after rebalancing.
        """
        node._height = 1 + max(_AVLNode.height(node._left), _AVLNode.height(node._right))
        balance = _AVLNode.height(node._left) - _AVLNode.height(node._right)
        
        if balance > 1:
            assert node._left is not None
            if _AVLNode.height(node._left._left) >= _AVLNode.height(node._left._right):
                node = _AVLNode._rotate_right(node)
            else:
                node._left = _AVLNode._rotate_left(node._left)
                node = _AVLNode._rotate_right(node)
        elif balance < -1:
            assert node._right is not None
            if _AVLNode.height(node._right._right) >= _AVLNode.height(node._right._left):
                node = _AVLNode._rotate_left(node)
            else:
                node._right = _AVLNode._rotate_right(node._right)
                node = _AVLNode._rotate_left(node)
        
        return node
    
    @staticmethod
    def _rotate_left(node: _AVLNode) -> _AVLNode:
        """
        Performs a left rotation on the given node.

        Parameters:
            node (_AVLNode): The root of the subtree to rotate.
        
        Returns:
            _AVLNode: The new root after rotation.
        """
        assert node._right is not None
        new_root = node._right
        node._right = new_root._left
        new_root._left = node
        
        node._height = 1 + max(_AVLNode.height(node._left), _AVLNode.height(node._right))
        new_root._height = 1 + max(_AVLNode.height(new_root._left), _AVLNode.height(new_root._right))
        
        return new_root
    
    @staticmethod
    def _rotate_right(node: _AVLNode) -> _AVLNode:
        """
        Performs a right rotation on the given node.

        Parameters:
            node (_AVLNode): The root of the subtree to rotate.
        
        Returns:
            _AVLNode: The new root after rotation.
        """
        assert node._left is not None
        new_root = node._left
        node._left = new_root._right
        new_root._right = node
        
        node._height = 1 + max(_AVLNode.height(node._left), _AVLNode.height(node._right))
        new_root._height = 1 + max(_AVLNode.height(new_root._left), _AVLNode.height(new_root._right))
        
        return new_root